pinot-controller/src/main/java/org/apache/pinot/controller/util/ConsumingSegmentInfoReader.java

```diff
@@ -131,6 +134,51 @@ private String generateServerURL(String tableNameWithType, String endpoint) {
     return String.format("%s/tables/%s/consumingSegmentsInfo", endpoint, tableNameWithType);
   }
 
+  /**
+   * Utility method to derive ingestion status from consuming segment Info. Status is HEALTHY if
+   * consuming segment info specifies CONSUMING state for all active segments across all servers
+   * including replicas.
+   */
+  public TableStatus.IngestionStatus getIngestionStatus(String tableNameWithType, int timeoutMs) {
+    try {
+      ConsumingSegmentsInfoMap consumingSegmentsInfoMap = getConsumingSegmentsInfo(tableNameWithType, timeoutMs);
+      for (Map.Entry<String, List<ConsumingSegmentInfo>> consumingSegmentInfoEntry : consumingSegmentsInfoMap._segmentToConsumingInfoMap
+          .entrySet()) {
+        String segmentName = consumingSegmentInfoEntry.getKey();
+        List<ConsumingSegmentInfo> consumingSegmentInfoList = consumingSegmentInfoEntry.getValue();
+        if (consumingSegmentInfoList == null || consumingSegmentInfoList.isEmpty()) {
+          String errorMessage = "Did not get any response from servers for segment: " + segmentName;
+          return TableStatus.IngestionStatus.newIngestionStatus(TableStatus.IngestionState.UNHEALTHY, errorMessage);
+        }
+
+        // Check if any responses are missing
+        Set<String> serversForSegment = _pinotHelixResourceManager.getServersForSegment(tableNameWithType, segmentName);
+        if (serversForSegment.size() != consumingSegmentInfoList.size()) {
+          Set<String> serversResponded =
+              consumingSegmentInfoList.stream().map(c -> c._serverName).collect(Collectors.toSet());
+          serversForSegment.removeAll(serversResponded);
+          String errorMessage =
+              "Not all servers responded for segment: " + segmentName + " Missing servers : " + serversForSegment;
+          return TableStatus.IngestionStatus.newIngestionStatus(TableStatus.IngestionState.UNHEALTHY, errorMessage);
+        }
+
+        for (ConsumingSegmentInfo consumingSegmentInfo : consumingSegmentInfoList) {
+          if (consumingSegmentInfo._consumerState
+              .equals(ConsumerState.NOT_CONSUMING.toString())) {
```

Comment:
We stop consumption during segment generation. Will that be treated as error?

Reply:
No that should be fine. the NOT_CONSUMING state is set when the LLRealtimeSegmentDataManager.State is set to ERROR.

Reply:
This means that if for some reason we have a transient failure with Kafka that results in CONSUMING segments going OFFLINE, we could be displaying HEALTHY for a long time.

We have had multiple combination of such issues at Linkedin, and over time, have arrived at the conclusion that monitoring the consumption metric (we emit 1 when consuming and 0 when not), is really the best alternative to decide whether or not some manual intervention is needed. A persistent value of 0 (where the time of persistence depends on application tolerance to non-fresh data) indicates call to action.

We can also check with how Uber handles this, since they are another large installation of real-time Pinot.

Reply:
From my read, it seemed like whenever ServerGauge.LLC_PARTITION_CONSUMING is set to 0 under error conditions (except for when consumption is complete or we're going from consuming to ONLINE), the 
LLRealtimeSegmentDataManager.State is set to ERROR. So it should work for any such erroneous conditions.

If that's not the case, then we can treat it as part of a bug fix for #6322 

In Uber, AFAIK we rely on the ingestion time lag (from Kafka) to determine problems in ingestions.

Reply:
We now include the freshness metric as well (as in how old is the data).

Linkedin Kafka drivers stopped exporting the per-partition lag metrics.

Reply:
+1 to Chinmay's observation. Hence when transient errors will make a segment OFFLINE in IS, it will be ERROR state in EV, and hence NOT_CONSUMING, resulting in UNHEALTHY.

Reply:
@npawar , @icefury71 when we change the state to OFFLINE in IS, EV will also be OFFLINE (not ERROR).

Reply:
regardless, the consumer will be in error state, and that is what the underlying server API is looking for